#7 Feat : 회원가입 - 아이디 중복 확인

uri010 opened the pull request:
## 요약
- 아이디 중복 확인

## 변경 내용
- 아이디 중복 확인 요청시 보낸 userId값과 중복되는 값이 user 테이블에 존재하는 개수를 조회해 그 값이 0이면 "ok"를, 0이 아니면 "존재하는 아이디입니다."를 반환합니다.
- 회원가입 요청시 아이디 중복 확인 후 존재하는 아이디가 아닐 경우만 회원가입이 가능하게 해줍니다. 

## Issue number/URL
#7 Feat : 회원가입 - 아이디 중복 확인

uri010 committed:
#7 feat : 아이디 중복 확인

요청으로 보낸 userId가 user 테이블에 존재하는 아이디인지 확인해서 중복이면 "존재하는 아이디입니다."를, 중복이 아니면 "ok"를 리턴

uri010 committed:
#7 feat : 회원가입 요청시 아이디 중복 확인 후 등록

회원정보를 등록하기 전에 아이디 중복을 확인하고 중복된 아이디가 없을 때만 회원을 등록

sunwootest commented on src/main/java/study/sharedcalendar/controller/UserController.java:
```diff
@@ -13,8 +13,16 @@ public class UserController {
     public String signUp(@RequestBody User user) {
-        userService.signUp(user);
-        return "ok";
+        if( userService.signUp(user))
```
if문 안에 불필요한 공백이 보이네요

sunwootest commented on src/main/java/study/sharedcalendar/controller/UserController.java:
```diff
@@ -13,8 +13,16 @@ public class UserController {
+        else return "존재하는 아이디입니다.";
+    }
+
+    @GetMapping("")
```
("")는 생략해도됩니다.

sunwootest commented on src/main/java/study/sharedcalendar/controller/UserController.java:
```diff
@@ -13,8 +13,16 @@ public class UserController {
+    @GetMapping("")
+    public String idCheck(@RequestParam String userId) {
+        int result = userService.idCheck(userId);
+        if( result == 0 ) return "ok";
+        else return "존재하는 아이디입니다.";
```
string을 return해서 메시지를 전달하는 방법은 지양해야합니다. 다른 방법이 어떻게 존재하는지 확인해보세요. 중복되는 아이디가 입력되었는데도 200 상태코드를 전달하네요.

sunwootest commented on src/main/java/study/sharedcalendar/service/UserService.java:
```diff
@@ -10,8 +10,17 @@
-        int result = userMapper.createUser(user);
+    public boolean signUp(User user) {
+        int isCheckedId = userMapper.idCheck(user.getUserId());
+        if( isCheckedId == 0 ) {
```
0보다는 false로 나타내는 것이 나아보입니다.

sunwootest commented on src/main/java/study/sharedcalendar/service/UserService.java:
```diff
@@ -10,8 +10,17 @@
+        else return false;
+    }
+
+    public int idCheck(String userId) {
```
int 보다는 boolean을 return 하는 것이 바람직해보이네요.

uri010 committed:
Merge branch 'feature/3' into feature/7

# Conflicts:
#	src/main/java/study/sharedcalendar/controller/UserController.java
#	src/main/java/study/sharedcalendar/mapper/UserMapper.java
#	src/main/java/study/sharedcalendar/service/UserService.java

uri010 committed:
#7 fix : 아이디 중복 확인 여부에 따른 상태 코드 수정

아이디 중복을 확인해서 중복이 없으면 상태코드 200을 보내고 중복이 있는 경우는 해당 DB 테이블에 같은 값이 있다는 의미로 상태코드 409를 전송
 회원가입시에도 로그인 중복 확인 절차를 한번 더 수행

uri010 commented:
리뷰 사항 반영했습니다.

- 아이디 중복 확인 시 결과에 따라 상태코드 수정
   - 중복이 없는 경우 : 상태코드 `200 OK`
   - 중복이 있는 경우 : 상태코드 `409 Conflict`
- 메소드 리턴 값 변경

sunwootest commented on src/main/java/study/sharedcalendar/controller/UserController.java:
```diff
@@ -11,8 +13,21 @@
-    @PostMapping("")
-    public void signUp(@RequestBody User user) {
-        userService.signUp(user);
+    @PostMapping
+    public ResponseEntity signUp(@RequestBody User user) {
+        if (userService.signUp(user)) {
+            return new ResponseEntity(HttpStatus.OK);
+        } else {
+            return new ResponseEntity(HttpStatus.CONFLICT);
+        }
+    }
```
클라이언트에서 특정 http 상태코드를 전송하기 위해 ResponseEntity를 반환했는데요, 스프링에서는 이를 좀 더 간편하게 적용하는 다른 방법도 존재하고, 예외 상황의 경우 ExceptionHandler를 이용해서 범용적으로 ResponseEntity와 함께 사용할 수도 있습니다. 지금은 괜찮지만 이렇게 코드를 짜다보면 Presentation Layer에 비즈니스 로직이 담길 수 밖에 없게 되고 코드가 지저분해집니다.

uri010 replied:
혹시 특정 http 상태코드 전송을 더 간편하게 적용하는 방법이 어노테이션 방식인가요?? 성공했을 때의 상태코드를 어노테이션으로 지정해놓고 실패했을 때는 ExceptionHandler를 이용하여 처리해주는 방식을 말하는 건지 확인 부탁드립니다. 

아래는 ExceptionHandler에 대해 참고하고자 하는 블로그입니다.
https://samtao.tistory.com/42

sunwootest replied:
그렇게도 적용할 수 있겠네요! 저는 현 코드에서 개선방향을 제안한 것이고 적절한 기능을 선택하는 것은 유리님 몫입니다! 그 기능을 사용할 때 합당한 이유만 있으면 됩니다!

sunwootest commented on src/main/java/study/sharedcalendar/service/UserService.java:
```diff
@@ -10,7 +10,20 @@
 
-    public void signUp(User user) {
-        userMapper.createUser(user);
+    public boolean signUp(User user) {
+        if (idCheck(user.getUserId())) {
+            userMapper.createUser(user);
+            return true;
+        }
+        else {
+            return false;
+        }
+    }
```
회원가입 메소드의 return 값을 boolean으로 처리해야해서 결국 controller 단에서 이에대한 응답을 다시 처리해줘야합니다. ExceptionHandler를 사용하면 회원가입 실패에 대한 상황에 ResponseEntity를 한번에 전송할 수도 있습니다.

sunwootest replied:
아래 idCheck 메소드도 마찬가지네요 :)

uri010 replied:
더 공부해서 반영하겠습니다!

uri010 committed:
#7 feat : 예외처리를 위한 ExceptionHandler 추가

uri010 committed:
#7 feat : 회원가입과 아이디 중복 확인 여부에 따른 예외 처리 추가

아이디가 중복이면 ID_DUPLICATE(409 Conflict) 예외 발생
회원가입시 잘못된 요청이 오면 INTERNAL_SERVER_ERROR(400 Bad Request)를 발생시켜 콘솔 에러 단순화

uri010 committed:
#7 fix : 코딩 컨벤션 준수

uri010 commented:
리뷰 사항 반영했습니다.

- 예외처리를 위한 ExceptionHandler 추가 
    - ErrorCode : status 코드와 그에 맞는 message를 enum 클래스로 관리합니다. 
    - ErrorResponse : ResponseEntity로 ErrorCode의 status와 message를 담는 객체입니다. 
    - CustomException : 예외를 던지기 위한 클래스로 따로 정의해준 예외를 처리할 때 사용합니다. 
    - GlobalExceptionHandler : 예외처리를 위한 handler로 CustomException의 ErrorCode에 따라 에러 메시지를 출력합니다. 
- UserService와 UserController 수정
    - 회원가입 : 아이디 중복 확인이 되면 회원가입을 수행합니다. 클라이언트 측에서 잘못된 요청을 보낸 경우 콘솔의 에러 메시지를 없애기 위해 INTERNAL_SERVER_ERROR 예외를 따로 처리해 줍니다.
    - 아이디 중복 확인 : 아이디 중복 확인 SQL문의 결과가 0이 아니라면 ID_DUPLICATE 예외로 처리해줍니다.

sunwootest commented on src/main/java/study/constant/ErrorCode.java:
```diff
@@ -0,0 +1,16 @@
+@Getter
+@AllArgsConstructor
+public enum ErrorCode {
+    INTERNAL_SERVER_ERROR(400, "파라미터 값을 확인해주세요."),
+
+    /* 409 CONFLICT */
+    ID_DUPLICATE(409, "중복된 아이디입니다");
```
409, 400과 같은 상수를 직접 사용하는 것은 지양해야합니다. 나은 대안이 있으면 그걸 적용해보세요 :)

uri010 replied:
HttpStatus의 상수를 이용해서 수정하겠습니다.

sunwootest commented on src/main/java/study/sharedcalendar/exception/CustomException.java:
```diff
@@ -0,0 +1,12 @@
+
+@Getter
+@RequiredArgsConstructor
+public class CustomException extends RuntimeException {
+    private final ErrorCode errorCode;
+}
```
약간의 피드백을 드리자면 RuntimeException보다 좀 더 구체적인 Exception을 상속하는 편이 낫습니다. 사실 대부분의 예외상황은 jdk에서 제공하는 Exception으로도 처리가 가능하거든요. 또 CustomException이라는 이름이 좀 모호하긴 하네요:)

uri010 replied:
더 자세한 Exception을 상속하고 CustomException이라는 이름 대신에 예외처리해주고자 하는 구체적인 예외 이름으로 바꾸게 되면 예외를 처리하는 클래스들이 너무 많아질 것 같습니다. 이렇게 구성해도 괜찮은건가요?

sunwootest replied:
예외를 구분하다 보면 종류 자체는 많지 않을겁니다. 예외의 파라미터로 받을 메시지가 많은거지 절대 예외의 종류가 많은건 아니거든요. 이 관점으로 한번 고민해보면 좋을 것 같습니다 :)

sunwootest replied:
일단 자바에서 제공하는 기본 예외부터 하나씩 알아보면 도움될겁니다!

sunwootest commented on src/main/java/study/sharedcalendar/mapper/UserMapper.java:
```diff
@@ -6,4 +6,7 @@
 public interface UserMapper {
     int createUser(User user);
+
+    int idCheck(String userId);
+
```
id 존재유무를 확인하는 것 같은데 sql에 종속적인 메소드를 작성한 느낌입니다. 의미상 boolean을 사용해야 맞을 것 같은데요. mysql에는 EXISTS라는 문법이 존재하는데 h2는 어떻게 처리해야할지 고민해보시면 좋을 것 같습니다. 참고로 지금은 임시로 h2를 사용하는 것이지 나중에는 다른 DBMS를 사용하는 편이 낫습니다 또 DBMS를 선택한 이유도 타당해야합니다. 만약에 나중에 h2를 선택한 이유가 무엇인가요라고  누군가 물었을 때 mysql 설치오류 때문이에요 라고 할 수는 없으니까요 :)

sunwootest commented on src/main/java/study/sharedcalendar/service/UserService.java:
```diff
@@ -3,14 +3,28 @@
+
+        try {
+            userMapper.createUser(user);
+        } catch (Exception e) {
+            throw new CustomException(INTERNAL_SERVER_ERROR);
+        }
```
예외는 목적이 뚜렷해야합니다. 제가 판단하기로는 Mapper에서 오류가 발생할 내가 예측하지 못한 예외상황을 하나로 묶어서 CustomException으로 예외전환을 한 것 같은데요, Exception으로 catch하는 것은 지양해야합니다.

sunwootest commented on src/main/java/study/sharedcalendar/service/UserService.java:
```diff
@@ -3,14 +3,28 @@
+        }
+    }
+
+    public void idCheck(String userId) {
+        if (userMapper.idCheck(userId) != 0)
+            throw new CustomException(ID_DUPLICATE);
     }
```
이 메소드는 재사용성이 뚜렷해보이는데 굳이 예외를 던져줄 필요가 없어보입니다. 회원가입 전용 아이디중복체크 메소드를 만들거나 컨트롤러 단에서 처리하는 편이 나아보이네요.

uri010 replied:
해당 메소드를 boolean으로 처리하고 컨트롤러에서 아이디 중복 확인을 하는 메소드에서 예외를 던지라는 의미인가요??

sunwootest reviewed (commented):
ExcpetionHandler를 잘 적용하셨네요. 코드리뷰 해드린 내용 반영하면 금방 속도 붙을 것 같습니다 :)

uri010 committed:
#7 fix : 예외처리 수정 및 아이디 중복 확인 SQL 문 수정

중복된 데이터 예외처리를 위해 DuplicatedException 클래스 생성
ErrorCode의 status 값의 타입을 HttpStatus로 수정
idCheck 메소드 이름을 idExist로 수정하고 리턴 타입을 boolean으로 수정

uri010 commented:
리뷰 사항 일부 반영했습니다.

- 중복된 데이터 예외처리를 위해 DuplicatedException 클래스를 생성합니다.
- ErrorCode의 status 값의 타입을 HttpStatus로 수정합니다.
- idCheck 메소드 이름을 idExist로 수정하고 리턴 타입을 boolean으로 수정합니다.
- UserMapper.xml의 idExist SQL 문에 exists 구문을 추가해 아이디 중복 여부를 boolean 값으로 반환합니다.

sunwootest commented on src/main/java/study/sharedcalendar/service/UserService.java:
```diff
@@ -3,14 +3,24 @@
+            userMapper.createUser(user);
+    }
+
+    public boolean idExist(String userId) {
+        return userMapper.idExist(userId);
     }
```
id와 userId는 다른 것 같은데 메소드명을 수정하는게 나아보입니다.

sunwootest commented on src/main/java/study/sharedcalendar/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,15 @@
+
+    @ExceptionHandler(DuplicateException.class)
+    protected ResponseEntity handleDuplicateException(DuplicateException ex) {
+        return ResponseEntity.status(ex.getErrorCode().getStatus()).body(new ErrorResponse(ex.getErrorCode().getStatus(), ex.getErrorCode().getMessage()));
```
코드가 길어지면 .이나 , 기준으로 개행을 해도 괜찮습니다.

sunwootest commented on src/main/java/study/sharedcalendar/controller/UserController.java:
```diff
@@ -11,8 +13,15 @@
-    @PostMapping("")
+    @PostMapping
     public void signUp(@RequestBody User user) {
+        System.out.println(user.getUserId());
         userService.signUp(user);
```
로그를 Sout으로 남기셨네요 :)

uri010 replied:
아 이거 테스트 용으로 넣어놨다가 깜빡하고 안지웠습니다 😅

sunwootest reviewed (commented):
리뷰반영 잘하셨네요:) 약간만 수정하시면 임시 approve 드리겠습니다!

uri010 committed:
#7 fix : idExist 메소드 명을 userIdExist로 수정 및 코드 컨벤션 준수

uri010 commented:
반영했습니다!

sunwootest commented:
임시 approve 드립니다 :) sout으로 테스트하는 습관을 버리시고 디버깅을 적용해보세요 훨씬 효율적입니다 :)

uri010 committed:
#7 feat : 아이디 중복 체크를 위해 UserService에  userIdDuplicationCheck 메소드 추가

Controller에 있는 비즈니스 로직을 없애줌
